### PR TITLE
crimson/osd/object_context_loader: Fix obc cache existence usage

### PR DESCRIPTION
**Pull request by Matan-B**
with_head_obc() uses get_cached_obc() to get_or_create obc instances.

If the obc exists in cache, get_or_load_obc is called with `existed`=true.
The assumption above is wrong.
Cache existence (`existed`) only guarantees that the obc instance was created (and inserted) in the obc_registery. However, it does **not** assure that the obc was actually loaded.

As obc-loading is now concurrent, it's possible for the first user to only create the obc in cache (without loading yet) and the second concurrent user to assume it was already loaded.

With this patch, we verify that the obc was loaded for get_or_load_obc usage.

* make loading-obc concurrent PR: https://github.com/ceph/ceph/pull/55488

Fixes: https://tracker.ceph.com/issues/64206
Fixes: https://tracker.ceph.com/issues/66214

---

```
  // obc loading is a concurrent phase. In case this obc is being loaded,
  // make other useres of this obc to await for the loading to complete.
```

Since we now await for loading to finish (in-case in progress), we can
also assert is_loaded().

Fixes: https://tracker.ceph.com/issues/65451

--- 
Fixes: https://tracker.ceph.com/issues/66308

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by athanatos — APPROVED** [review]

**Review by xxhdx1985126 — APPROVED** [review]

File: src/crimson/osd/object_context_loader.cc
Comment (cyx1231st) [inline src/crimson/osd/object_context_loader.cc:154]: I think this is the correct direction to return only until the `obc` is loaded.

> As obc-loading is now concurrent, it's possible for the first user to only create the obc in cache (without loading yet) and the second concurrent user to assume it was already loaded.

But I think the first user must create and also start to load the `obc` at the same time. So the second (maybe-concurrent) user should assume it is already loading or already loaded.

(If I'm wrong above, can you help point to the violating logic?)

If I'm correct above, I think the accurate pseudo logic might be:
```c++
if (existed && obc->is_loaded()) {
  return load_obc_iertr::make_ready_future<ObjectContextRef>(obc);
} else if (existed) {
  // not the first user, the obc must be already loading,
  // return the loading future, instead of loading the obc again.
} else { // !existed: the first user
  return obc->template with_promoted_lock<State, IOInterruptCondition>(
    [obc, this] { return load_obc(obc); }
  );
}
```

However, the current version should be fine to address the assert failure.

Comment (Matan-B) [reply src/crimson/osd/object_context_loader.cc:154]: >But I think the first user must create and also start to load the obc at the same time. So the second (maybe-concurrent) user should assume it is already loading or already loaded.

This is correct. Broadly speaking, the pseudo logic should work, though I'm now sure it's worth the additional logic. This will require maintaining the loading future from the "first-user" for the "second" one to use later, instead of loading again.
The main reason for avoiding implementing the pseudo logic is that this scenario is rare. This was quite frequent in testing was because the obc_lru_size is set to 10 (to reduce cache hits).

@cyx1231st, I suggest merging as is to resolve the bug. Later on, we could rework this bit if needed. What do you think?

Comment (xxhdx1985126) [reply src/crimson/osd/object_context_loader.cc:154]: > This is correct. Broadly speaking, the pseudo logic should work, though I'm now sure it's worth the additional logic. This will require maintaining the loading future from the "first-user" for the "second" one to use later, instead of loading again.


@Matan-B https://github.com/ceph/ceph/pull/57687 is for this purpose, it avoid loading the obc again by making sure the `exist` check and the load happens synchronously, maybe we should consider it?

Comment (Matan-B) [reply src/crimson/osd/object_context_loader.cc:154]: > @Matan-B #57687 is for this purpose, it avoid loading the obc again by making sure the `exist` check and the load happens synchronously, maybe we should consider it?

@xxhdx1985126 , To consider the benefit of avoiding the extra load we need to understand how frequent this scenario is. IIUC, a normal workload (not thrash test), with a decent size of obc_lru_size (default of 512) - should not be impacted by this.
For this reason, I'm leaning towards readability and maintainability here.

Comment (Matan-B) [reply src/crimson/osd/object_context_loader.cc:154]: Note: I'm scheduling a test run since this assert causes too many jobs to fail and merging is blocked, please let me know if you don't agree with my previous comments.

Comment (xxhdx1985126) [reply src/crimson/osd/object_context_loader.cc:154]: @Matan-B @cyx1231st I think this pr may fail. According to the following code, we can't promote the lock to EXCL when there are multiple readers/writers. Am I missing something?

```
void tri_mutex::promote_from_read()    
{                                      
  assert(readers == 1);                
  --readers;                           
  exclusively_used = true;             
}                                      

void tri_mutex::promote_from_write()    
{                                       
  assert(writers == 1);                 
  --writers;                            
  exclusively_used = true;              
}                                       
```

Comment (cyx1231st) [reply src/crimson/osd/object_context_loader.cc:154]: > @cyx1231st, I suggest merging as is to resolve the bug. Later on, we could rework this bit if needed. What do you think?

I'm fine to address the specific assert failure with a simple fix, and then find a proper way to avoid loading the same obc.

And I think it is also related to https://tracker.ceph.com/issues/65451, because allowing the non-first user(s) to load the same obc (in the promoted/exclusive lock) should violate the assertion `readers/writers == 1' upon promotion.

> This will require maintaining the loading future from the "first-user" for the "second" one to use later, instead of loading again. The main reason for avoiding implementing the pseudo logic is that this scenario is rare. ..

A per-obc `seastar::shared_mutex<>` may be helpful to make the concurrent users (except the first user) to wait outside the promoted/exclusive lock until the obc is loaded, and also preserve the waiting order. Otherwise, we might need to rethink tri-mutex and the lock promotion. Anyway, the exclusive constraint to load the same obc makes more sense to me.

File: src/crimson/osd/object_context_loader.cc
Comment (xxhdx1985126) [inline src/crimson/osd/object_context_loader.cc:152]: seastar::with_shared?

Comment (xxhdx1985126) [reply src/crimson/osd/object_context_loader.cc:152]: @cyx1231st Maybe SeaStore's `Cache::get_extent()`-like methods can also follow this approach?

**Comment by Matan-B** [comment]
Approved commits left intact, I'll squash after CR.

**Review by xxhdx1985126 — APPROVED** [review]

**Review by cyx1231st — APPROVED** [review]

File: src/crimson/osd/object_context_loader.cc
Comment (cyx1231st) [reply src/crimson/osd/object_context_loader.cc]: @xxhdx1985126 Sorry I'm not following. Which mutex/lock are you refering to in the `Cache::get_extent()` related path?

Comment (xxhdx1985126) [reply src/crimson/osd/object_context_loader.cc]: > @xxhdx1985126 Sorry I'm not following. Which mutex/lock are you refering to in the `Cache::get_extent()` related path?

Sorry, I didn't make myself clear. With regard to `SeaStore::Cache`, we have a similar issue and we need to make sure that the check for whether an extent is in `Cache` and the load of the missing extent are synchronous. Maybe we can take an approach like this pr, adding an mutex/lock for each extent and lock it when trying to load the extent from disk. What do you think?

Comment (cyx1231st) [reply src/crimson/osd/object_context_loader.cc]: If there is no specific blocking problem that the current way cannot cover (or it result in unacceptabe complexity). Probably it's better to keep the existing way.

The mutex added in this case is because otherwise the code will break.

**Comment by Matan-B** [comment]
changeset: with_lock -> interruptor.with_lock

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-05-30_07:12:43-crimson-rados-wip-matanb-crimson-testing-may-28-v3-distro-crimson-smithi/7734043

OSD2
```
DEBUG 2024-05-30 08:12:53,141 [shard 0:main] osd -  pg_epoch 16 pg[3.6( v 16'96 (0'0,16'96] local-lis/les=14/15 n=12 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [2,3,1] r=0 lpr=14 luod=16'89 lua=0'0 crt=16'96 lcod 16'89 mlcod 16'89 active+clean  ObjectContextLoader::with_head_obc: object 3:6ec6ac80:::benchmark_data_smithi007_34563_object386:head
DEBUG 2024-05-30 08:12:53,141 [shard 0:main] osd -  pg_epoch 16 pg[3.6( v 16'96 (0'0,16'96] local-lis/les=14/15 n=12 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [2,3,1] r=0 lpr=14 luod=16'89 lua=0'0 crt=16'96 lcod 16'89 mlcod 16'89 active+clean  ObjectContextLoader::get_or_load_obc: 3:6ec6ac80:::benchmark_data_smithi007_34563_object386:head -- fully_loaded=false, invalidated_by_interval_change=false
DEBUG 2024-05-30 08:12:53,142 [shard 0:main] osd -  pg_epoch 16 pg[3.6( v 16'96 (0'0,16'96] local-lis/les=14/15 n=12 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [2,3,1] r=0 lpr=14 luod=16'90 lua=0'0 crt=16'96 lcod 16'90 mlcod 16'90 active+clean  ClientRequest::process_op: client_request(id=16778573, detail=m=[osd_op(client.4438.0:3090 3.6 3:6ec6ac80:::benchmark_data_smithi007_34563_object386:head {set-alloc-hint object_size 65536 write_size 8192, write 8192~8192 in=8192b} snapc 0={} ondisk+write+known_if_redirected+supports_pool_eio e16)]).0: entered get_obc stage, about to wait_scrub
DEBUG 2024-05-30 08:12:53,142 [shard 0:main] osd -  pg_epoch 16 pg[3.6( v 16'96 (0'0,16'96] local-lis/les=14/15 n=12 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [2,3,1] r=0 lpr=14 luod=16'90 lua=0'0 crt=16'96 lcod 16'90 mlcod 16'90 active+clean  ClientRequest::process_op: client_request(id=16778573, detail=m=[osd_op(client.4438.0:3090 3.6 3:6ec6ac80:::benchmark_data_smithi007_34563_object386:head {set-alloc-hint object_size 65536 write_size 8192, write 8192~8192 in=8192b} snapc 0={} ondisk+write+known_if_redirected+supports_pool_eio e16)]).0: past scrub blocker, getting obc
DEBUG 2024-05-30 08:12:53,142 [shard 0:main] osd -  pg_epoch 16 pg[3.6( v 16'96 (0'0,16'96] local-lis/les=14/15 n=12 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [2,3,1] r=0 lpr=14 luod=16'90 lua=0'0 crt=16'96 lcod 16'90 mlcod 16'90 active+clean  ObjectContextLoader::with_head_obc: object 3:6ec6ac80:::benchmark_data_smithi007_34563_object386:head
DEBUG 2024-05-30 08:12:53,142 [shard 0:main] osd -  pg_epoch 16 pg[3.6( v 16'96 (0'0,16'96] local-lis/les=14/15 n=12 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [2,3,1] r=0 lpr=14 luod=16'90 lua=0'0 crt=16'96 lcod 16'90 mlcod 16'90 active+clean  ObjectContextLoader::get_or_load_obc: 3:6ec6ac80:::benchmark_data_smithi007_34563_object386:head -- fully_loaded=false, invalidated_by_interval_change=false
DEBUG 2024-05-30 08:12:53,142 [shard 0:main] osd -  pg_epoch 16 pg[3.6( v 16'96 (0'0,16'96] local-lis/les=14/15 n=12 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [2,3,1] r=0 lpr=14 luod=16'90 lua=0'0 crt=16'96 lcod 16'90 mlcod 16'90 active+clean  ObjectContextLoader::get_or_load_obc: cache miss on 3:6ec6ac80:::benchmark_data_smithi007_34563_object386:head

ceph-osd: /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-3821-gb37feebb/rpm/el9/BUILD/ceph-19.0.0-3821-gb37feebb/src/crimson/common/tri_mutex.cc:134: void tri_mutex::promote_from_write(): Assertion `writers == 1' failed.
```

@cyx1231st, this looks like 2 users entered `get_or_load_obc`. IIUC, `obc->with_lock` should have not allow that based on the fact that both users are `lock.for_write()`, right?

**Comment by cyx1231st** [comment]
Yeah, the per-obc mutex should have made loading exclusive. That looks unexpected.

**Comment by Matan-B** [comment]
> Yeah, the per-obc mutex should have made loading exclusive. That looks unexpected.

I was able to pin down the cause:
```
DEBUG 2024-06-02 15:43:08,275 [shard 1:main] osd - tri_mutex benchmark_data_folio01_1458381_object41 writers 0 readers 0 exclusively_used false waiters: 0 tri_mutex::lock_for_write(): 
DEBUG 2024-06-02 15:43:08,275 [shard 1:main] osd - tri_mutex benchmark_data_folio01_1458381_object41 writers 0 readers 0 exclusively_used false waiters: 0 tri_mutex::try_lock_for_write(): 
DEBUG 2024-06-02 15:43:08,275 [shard 1:main] osd - tri_mutex benchmark_data_folio01_1458381_object41 writers 1 readers 0 exclusively_used false waiters: 0 tri_mutex::lock_for_write(): lock_for_write successfully
DEBUG 2024-06-02 15:43:08,276 [shard 1:main] osd - tri_mutex benchmark_data_folio01_1458381_object41 writers 1 readers 0 exclusively_used false waiters: 0 tri_mutex::lock_for_write(): 
DEBUG 2024-06-02 15:43:08,276 [shard 1:main] osd - tri_mutex benchmark_data_folio01_1458DEBUG 2024-06-02 15:43:08,276 [shard 1:main] osd - tri_mutex benchmark_data_folio01_1458381_object41 writers 1 readers 0 exclusively_used false waiters: 0 tri_mutex::try_lock_for_write(): 
DEBUG 2024-06-02 15:43:08,276 [shard 1:main] osd - tri_mutex benchmark_data_folio01_1458381_object41 writers 2 readers 0 exclusively_used false waiters: 0 tri_mutex::lock_for_write(): lock_for_write successfully
DEBUG 2024-06-02 15:43:08,277 [shard 1:main] osd - tri_mutex benchmark_data_folio01_1458381_object41 writers 2 readers 0 exclusively_used false waiters: 0 tri_mutex::promote_from_write():
FAILED: assert(writers == 1);
```

@cyx1231st, @athanatos - We are allowing multiple write operations on the same `tri_mutex`. However, we expect `promote_from_write` to have a single writer at all times `assert(writers == 1)`.

File: src/crimson/common/tri_mutex.cc
Comment (Matan-B) [inline src/crimson/common/tri_mutex.cc:9]: I'll introduce this subsys once approved in order to avoid recompiling.

**Comment by cyx1231st** [comment]
@Matan-B Thanks for the investigation!

I see, the problem is that the first writer 0 enters the tri-mutex write lock, but it is not the one promoting the lock, it becomes the writer 2 instead.

The lock and promotion from the first writer should be atomic in order to pass the `assert(writers == 1)`, and the fix https://github.com/ceph/ceph/blob/f1761d6239665b769e2de9948cdea14bfd7facd2/src/crimson/osd/object_context.h#L141 from https://github.com/ceph/ceph/pull/56844 only tried to make the promotion atomic. Looking at the calling path, the `obc->with_lock()` still isn't atomic for the 1st locker.

**Comment by Matan-B** [comment]
> @Matan-B Thanks for the investigation!
> 
> I see, the problem is that the first writer 0 enters the tri-mutex write lock, but it is not the one promoting the lock, it becomes the writer 2 instead.
> 
> The lock and promotion from the first writer should be atomic in order to pass the `assert(writers == 1)`, and the fix
> 
> https://github.com/ceph/ceph/blob/f1761d6239665b769e2de9948cdea14bfd7facd2/src/crimson/osd/object_context.h#L141
> 
> from #56844 only tried to make the promotion atomic. Looking at the calling path, the `obc->with_lock()` still isn't atomic for the 1st locker.

Do you agree we should disallow pipelining writes for the moment?

**Comment by cyx1231st** [comment]
Probably https://github.com/ceph/ceph/pull/57825 can address the specific atomic issue. Feel free to include that PR into this.

**Comment by Matan-B** [comment]
> Probably #57825 can address the specific atomic issue. Feel free to include that PR into this.

@cyx1231st, included. Thanks!

**Comment by cyx1231st** [comment]
^ expecting https://tracker.ceph.com/issues/66308 and https://tracker.ceph.com/issues/65451 to be fixed.

File: src/crimson/osd/object_context_loader.cc
Comment (cyx1231st) [inline src/crimson/osd/object_context_loader.cc:21]: I think is intentional @tchaikov, this is because structured bindings don't declare variables, see https://www.reddit.com/r/cpp_questions/comments/e1ralf/why_structured_bindings_cant_declare_variables/

File: src/crimson/osd/object_context_loader.cc
Comment (cyx1231st) [inline src/crimson/osd/object_context_loader.cc:88]: Same here.

File: src/crimson/osd/object_context_loader.cc
Comment (cyx1231st) [inline src/crimson/osd/object_context_loader.cc:93]: nit, this is usually:
```
).safe_then_interruptible(
```

**Review by cyx1231st — APPROVED** [review]

**Comment by Matan-B** [comment]
It seems that 55e7e13109c7085486efbea21b02324a3bb90352 couldn't ensure the atomic call.
* The first user tries (16790888) to load:
`ObjectContextLoader::with_head_obc: object 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head existed false`
* with_lock called:
`tri_mutex benchmark_data_smithi120_40431_object722 writers 1 readers 0 exclusively_used false waiters: 0 tri_mutex::lock_for_write(): lock_for_write successfully`
* get_or_load called:
`ObjectContextLoader::get_or_load_obc: 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head -- fully_loaded=false, invalidated_by_interval_change=false`
* Second user tries (16790889) to load and **assumes** `existed=true` 
`ObjectContextLoader::with_head_obc: object 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head existed true`

Sanitized crash:
```
ClientRequest::process_op: client_request(id=16790889, detail=m=[osd_op(client.4840.0:5778 5.2 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head: not completed, entering get_obc stage
ClientRequest::process_op: client_request(id=16790888, detail=m=[osd_op(client.4840.0:5777 5.2 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head: past scrub blocker, getting obc
ObjectContextLoader::with_head_obc: object 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head existed false
tri_mutex benchmark_data_smithi120_40431_object722 writers 0 readers 0 exclusively_used false waiters: 0 tri_mutex::lock_for_write(): 
tri_mutex benchmark_data_smithi120_40431_object722 writers 0 readers 0 exclusively_used false waiters: 0 tri_mutex::try_lock_for_write(): 
tri_mutex benchmark_data_smithi120_40431_object722 writers 1 readers 0 exclusively_used false waiters: 0 tri_mutex::lock_for_write(): lock_for_write successfully
ObjectContextLoader::get_or_load_obc: 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head -- fully_loaded=false, invalidated_by_interval_change=false
ClientRequest::process_op: client_request(id=16790889, detail=m=[osd_op(client.4840.0:5778 5.2 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head: entered get_obc stage, about to wait_scrub
ClientRequest::process_op: client_request(id=16790889, detail=m=[osd_op(client.4840.0:5778 5.2 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head: past scrub blocker, getting obc
ObjectContextLoader::with_head_obc: object 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head existed true
tri_mutex benchmark_data_smithi120_40431_object722 writers 1 readers 0 exclusively_used false waiters: 0 tri_mutex::lock_for_write(): 
tri_mutex benchmark_data_smithi120_40431_object722 writers 1 readers 0 exclusively_used false waiters: 0 tri_mutex::try_lock_for_write(): 
tri_mutex benchmark_data_smithi120_40431_object722 writers 2 readers 0 exclusively_used false waiters: 0 tri_mutex::lock_for_write(): lock_for_write successfully
ObjectContextLoader::get_or_load_obc: 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head -- fully_loaded=false, invalidated_by_interval_change=false
ObjectContextLoader::get_or_load_obc: cache miss on 5:41d9d6fe:::benchmark_data_smithi120_40431_object722:head
tri_mutex benchmark_data_smithi120_40431_object722 writers 2 readers 0 exclusively_used false waiters: 0 tri_mutex::promote_from_write(): 

src/crimson/common/tri_mutex.cc:163: void tri_mutex::promote_from_write(): Assertion `writers == 1' failed.
```

CC: @cyx1231st 

https://pulpito.ceph.com/matan-2024-06-05_08:27:17-crimson-rados-wip-matanb-crimson-load-exist-distro-crimson-smithi/7742180/

**Comment by cyx1231st** [comment]
Quote my previous comment.

> A per-obc seastar::shared_mutex<> may be helpful to make the concurrent users (except the first user) to wait outside the promoted/exclusive lock until the obc is loaded, and also preserve the waiting order.

After a second thought, I think "preserve the waiting order" might not be a necessary requirement here, the design of phases (if implemented correctly) should already sufficient to recover the original order upon entering the next exclusive phase: https://github.com/ceph/ceph/blob/d18b722557af8fc082d578547fa7be07a8ffc49e/src/crimson/osd/osd_operations/client_request.cc#L347-L357
(i.e. upon entering the next exclusive "process" phase)

> Update: It looks that the call interruptor::with_lock makes the func that is passed to _with_lock to be non-atomic.

I agree with you, `seastar::shared_mutex::lock()` also has the same property that returns a future unconditionally, refering to  https://github.com/scylladb/seastar/blob/47bfd73e304fd02ed91d7edf6c60d720e767eac5/include/seastar/core/shared_mutex.hh#L109-L129. Probably we should use the atomic version `try_lock() -> bool` instead.

> Moreover, since with https://github.com/ceph/ceph/commit/55e7e13109c7085486efbea21b02324a3bb90352 the first user should load atomically - we may no longer need the per-obc mutex (https://github.com/ceph/ceph/commit/e456bd7cdd1ad1cd68026cebedd5b1d9d01001f7).
> Assuming the first user (aka loader) will call with_promoted_lock atomically.

I think the assumption is incorrect, see the promoted lock interface: https://github.com/ceph/ceph/blob/d18b722557af8fc082d578547fa7be07a8ffc49e/src/crimson/common/tri_mutex.h#L28-L39

I don't think these promoted locks support waiting yet, their `lock()` simply returns `void`. They are more like placeholders (partial implementations) only asserting that the promotion behavior should be correct. The waiting is still needed, either it is the per-obc mutex, or by making the promoted lock classes more complete by adding a similar mechanism inside.

**Comment by Matan-B** [comment]
> I think the assumption is incorrect, see the promoted lock interface:

> I don't think these promoted locks support waiting yet, their `lock()` simply returns `void`. They are more like placeholders (partial implementations) only asserting that the promotion behavior should be correct. The waiting is still needed, either it is the per-obc mutex, or by making the promoted lock classes more complete by adding a similar mechanism inside.

I have switched to the atmoic version when using the per-obc mutex.

I will note that on local tests I was not able to trigger the per-obc mutex to actually make any user wait. It looks to me that the call to promote was atomic and therefore the second user was waiting **outside** of get_or_load (thanks to `ObjectContext::_with_lock`). However, this might be simply luck and it's better to still have per-obc mutex to cover all cases (as rare as they might be).

**Comment by Matan-B** [comment]
@athanatos @cyx1231st, results look great!
https://pulpito.ceph.com/matan-2024-06-06_12:44:30-crimson-rados-wip-matanb-crimson-load-exist-distro-crimson-smithi/

I want to merge this asap in order to unblock other awaiting PRs, PTAL.

**Review by athanatos — APPROVED** [review]
LGTM

**Comment by mattbenjamin** [comment]
am I correct that a "tri_mutex" is not a mutex but rather a future-enabled reader-writer lock?

**Comment by Matan-B** [comment]
> am I correct that a "tri_mutex" is not a mutex but rather a future-enabled reader-writer lock?

Yes, though not a traditional reader-writer lock, attaching the header comment:
```
/// shared/exclusive mutual exclusion
///
/// Unlike reader/write lock, tri_mutex does not enforce the exclusive access
/// of write operations, on the contrary, multiple write operations are allowed
/// to hold the same tri_mutex at the same time. Here, what we mean is that we
/// can pipeline reads, and we can pipeline writes, but we cannot allow a read
/// while writes are in progress or a write while reads are in progress.
///
/// tri_mutex is based on seastar::shared_mutex, but instead of two kinds of
/// waiters, tri_mutex keeps track of three kinds of lock users:
/// - readers
/// - writers
/// - exclusive users
///
/// For lock promotion, a read or a write lock is only allowed to be promoted
/// atomically upon the first locking.
```

**Comment by athanatos** [comment]
@mattbenjamin Kinda, it's multi-writer, multi-reader, single-exclusive.  It's there to model the OSD's rules for dealing with concurrent operations on a single object.

**Pull request merged**
